File: iOS/Services/ActivationRetention.swift
import Foundation

public enum ActivationMilestone: String, CaseIterable, Hashable, Sendable {
    case clientReady = "client_ready"
    case routineReady = "routine_ready"
    case firstAction = "first_action"
    case reminderReady = "reminder_ready"
}

public struct ActivationProgress: Equatable, Sendable {
    public let completed: Set<ActivationMilestone>

    public var coreCompletedCount: Int {
        Self.coreMilestones.filter(completed.contains).count
    }

    public var firstValueReached: Bool {
        Self.coreMilestones.allSatisfy(completed.contains)
    }

    public var nextCoreMilestone: ActivationMilestone? {
        Self.coreMilestones.first { !completed.contains($0) }
    }

    public static let coreMilestones: [ActivationMilestone] = [
        .clientReady,
        .routineReady,
        .firstAction
    ]
}

public enum ActivationRetentionStore {
    public static func progress(defaults: UserDefaults = .standard) -> ActivationProgress {
        let completed = Set(ActivationMilestone.allCases.filter { defaults.bool(forKey: key($0)) })
        return ActivationProgress(completed: completed)
    }

    @discardableResult
    public static func mark(
        _ milestone: ActivationMilestone,
        defaults: UserDefaults = .standard
    ) -> Bool {
        guard !defaults.bool(forKey: key(milestone)) else { return false }
        defaults.set(true, forKey: key(milestone))
        DiagnosticsReporter.event(
            "activation_milestone",
            fields: ["milestone": milestone.rawValue, "state": "reached"]
        )
        return true
    }

    public static func reconcile(
        clientReady: Bool,
        routineReady: Bool,
        firstAction: Bool,
        reminderReady: Bool,
        defaults: UserDefaults = .standard
    ) -> ActivationProgress {
        if clientReady { mark(.clientReady, defaults: defaults) }
        if routineReady { mark(.routineReady, defaults: defaults) }
        if firstAction { mark(.firstAction, defaults: defaults) }
        if reminderReady { mark(.reminderReady, defaults: defaults) }
        return progress(defaults: defaults)
    }

    private static func key(_ milestone: ActivationMilestone) -> String {
        "oakActivation_\(milestone.rawValue)"
    }
}

File: Android/app/src/main/java/com/example/supplementtracker/service/ActivationRetention.kt
package com.example.supplementtracker.service

import android.content.Context
import android.content.SharedPreferences

enum class ActivationMilestone(val wireName: String) {
    CLIENT_READY("client_ready"),
    ROUTINE_READY("routine_ready"),
    FIRST_ACTION("first_action"),
    REMINDER_READY("reminder_ready")
}

data class ActivationProgress(val completed: Set<ActivationMilestone>) {
    val coreCompletedCount: Int
        get() = coreMilestones.count { it in completed }

    val firstValueReached: Boolean
        get() = coreMilestones.all { it in completed }

    val nextCoreMilestone: ActivationMilestone?
        get() = coreMilestones.firstOrNull { it !in completed }

    companion object {
        val coreMilestones = listOf(
            ActivationMilestone.CLIENT_READY,
            ActivationMilestone.ROUTINE_READY,
            ActivationMilestone.FIRST_ACTION
        )
    }
}

class ActivationRetentionStore(
    private val context: Context,
    private val prefs: SharedPreferences = OakPrefs.get(context)
) {
    fun progress(): ActivationProgress {
        val completed = ActivationMilestone.entries.filterTo(mutableSetOf()) { milestone ->
            prefs.getBoolean(key(milestone), false)
        }
        return ActivationProgress(completed)
    }

    fun mark(milestone: ActivationMilestone): Boolean {
        if (prefs.getBoolean(key(milestone), false)) return false
        prefs.edit().putBoolean(key(milestone), true).apply()
        DiagnosticsReporter.event(
            context,
            "activation_milestone",
            mapOf("milestone" to milestone.wireName, "state" to "reached")
        )
        return true
    }

    fun reconcile(
        clientReady: Boolean,
        routineReady: Boolean,
        firstAction: Boolean,
        reminderReady: Boolean
    ): ActivationProgress {
        if (clientReady) mark(ActivationMilestone.CLIENT_READY)
        if (routineReady) mark(ActivationMilestone.ROUTINE_READY)
        if (firstAction) mark(ActivationMilestone.FIRST_ACTION)
        if (reminderReady) mark(ActivationMilestone.REMINDER_READY)
        return progress()
    }

    private fun key(milestone: ActivationMilestone): String {
        return "oakActivation_${milestone.wireName}"
    }
}
